### ISSUE TITLE
Add https/basic auth/client certs possibilities

### ISSUE THREAD

**Issue by jhd85** [kevinkamps/NomadJobUpdater#1]
Hello!

I'm not sure if I can request features here. But I would like some new features for this tool.
- Support for a HTTPS URL combined with a client cert (which nomad can verify)
- Support for a basic auth on the URL

Maybe I can contribute, but I'm not very familiar with go. ;)

Thanks!

**Comment by kevinkamps** [comment]
Sure you can. I would be happy to help you out. I will try to look at it this weekend. 

Have you perhaps got a test setup that i can use to test this? Something i can put in front of my Nomad because i don't think Nomad supports these features. Maybe an NGINX or something like that in a docker container for example?

**Comment by jhd85** [comment]
For the nginx TLS part you could use a config like this:

```
server {
    listen              443 ssl;
    ssl_certificate     /data/nomad/certs/client.pem;
    ssl_certificate_key /data/nomad/certs/client-key.pem;

    server_name  nomad.drewes.nu;
    auth_basic           "Nomad administration";
    auth_basic_user_file /data/nomad/.htpasswd;

    location / {
      proxy_pass      https://10.25.10.15:4646;
      proxy_ssl_certificate /data/nomad/certs/server.pem;
      proxy_ssl_certificate_key /data/nomad/certs/server-key.pem;
      proxy_ssl_trusted_certificate /data/nomad/certs/nomad-ca.pem;
    }
}
```

For cert creation I have some ansible script:

```
#
# certificates
#

- hosts: localhost
  gather_facts: True

  vars_files:
    - ../Vars/Global.yml

  tasks:

  - name: Create tmp dir
    file:
      path: "{{ playbook_dir }}/../tmp/"
      state: directory
      mode: '0700'

  - name: Getting cfssl binaries
    command: wget {{ item.url }} -O {{ item.name }}
    with_items:
      - { url: "https://pkg.cfssl.org/R1.2/cfssl_linux-amd64", name: "{{ playbook_dir }}/../tmp/cfssl"}
      - { url: "https://pkg.cfssl.org/R1.2/cfssljson_linux-amd64", name: "{{ playbook_dir }}/../tmp/cfssljson"}

  - name: Change execute rights
    file:
      path: "{{ item }}"
      mode: '0770'
    with_items:
      - "{{ playbook_dir }}/../tmp/cfssl"
      - "{{ playbook_dir }}/../tmp/cfssljson"

  - name: Generate CA
    shell: "./cfssl print-defaults csr | ./cfssl gencert -initca - | ./cfssljson -bare nomad-ca"
    args:
      chdir: "{{ playbook_dir }}/../tmp/"

  - name: Copying TLS config file
    template: src='{{ playbook_dir }}/../Files/TLS/root/cfssl/cfssl.json' dest='{{ playbook_dir }}/../tmp/cfssl.json' mode=660 force=yes

  - name: Generate certs and keys
    shell: "{{ item }}"
    args:
      chdir: "{{ playbook_dir }}/../tmp/"
    with_items:
      - "echo '{}' | ./cfssl gencert -ca=nomad-ca.pem -ca-key=nomad-ca-key.pem -config={{ playbook_dir }}/../tmp/cfssl.json -hostname=\"server.global.nomad,10.25.10.15\" - | ./cfssljson -bare server"
      - "echo '{}' | ./cfssl gencert -ca=nomad-ca.pem -ca-key=nomad-ca-key.pem -config={{ playbook_dir }}/../tmp/cfssl.json -hostname=\"client.global.nomad\" - | ./cfssljson -bare client"
      - "echo '{}' | ./cfssl gencert -ca=nomad-ca.pem -ca-key=nomad-ca-key.pem -profile=client - | ./cfssljson -bare cli"

- hosts: Manager:NomadWorkerPriv:NomadWorkerDMZ
  become: yes

  tasks:
  - name: Create nomad data dir
    file:
      path: /data/nomad/certs/
      state: directory
      mode: '0755'

- hosts: Manager
  become: yes

  tasks:
  - name: Copy certificates
    copy:
      src: "{{ playbook_dir }}/../tmp/{{ item.name }}"
      dest: "/data/nomad/certs/{{ item.name }}"
      owner: root
      group: root
      mode: '0600'
    with_items:
    - { name: "nomad-ca.pem" }
    - { name: "nomad-ca-key.pem" }
    - { name: "server.pem" }
    - { name: "server-key.pem" }
    - { name: "client.pem" }
    - { name: "client-key.pem" }
    - { name: "cli.pem" }
    - { name: "cli-key.pem" }

  - name: Is nomad already running?
    shell: systemctl status nomad
    register: nomad_status
    ignore_errors: true

  - name: Restart nomad
    shell: systemctl restart nomad
    when: nomad_status is defined and not nomad_status.rc

- hosts: NomadWorkerPriv:NomadWorkerDMZ
  become: yes

  tasks:
  - name: Copy certificates
    copy:
      src: "{{ playbook_dir }}/../tmp/{{ item.name }}"
      dest: "/data/nomad/certs/{{ item.name }}"
      owner: root
      group: root
      mode: '0600'
    with_items:
    - { name: "nomad-ca.pem" }
    - { name: "client.pem" }
    - { name: "client-key.pem" }

  - name: Is nomad already running?
    shell: systemctl status nomad
    register: nomad_status
    ignore_errors: true

  - name: Restart nomad
    shell: systemctl restart nomad
    when: nomad_status is defined and not nomad_status.rc

- hosts: localhost

  tasks:
  - name: Remove certificates
    file:
      path: "{{ playbook_dir }}/../tmp/{{ item.name }}"
      state: absent
    with_items:
    - { name: "nomad-ca.pem" }
    - { name: "nomad-ca.csr" }
    - { name: "nomad-ca-key.pem" }
    - { name: "server.pem" }
    - { name: "server.csr" }
    - { name: "server-key.pem" }
    - { name: "client.pem" }
    - { name: "client.csr" }
    - { name: "client-key.pem" }
    - { name: "cli.pem" }
    - { name: "cli.csr" }
    - { name: "cli-key.pem" }
```
You can use whatever you want of course. 

The cfssl json looks like this:
```
{
  "signing": {
    "default": {
      "expiry": "87600h",
      "usages": ["signing", "key encipherment", "server auth", "client auth"]
    }
  }
}
```

Just update the nomad config with:
```
tls {
  http = true
  rpc  = true

  ca_file   = "/data/nomad/certs/nomad-ca.pem"
  cert_file = "/data/nomad/certs/server.pem"
  key_file  = "/data/nomad/certs/server-key.pem"

  verify_server_hostname = false
  verify_https_client    = true
}
```
Same for the workers but use the client cert.

**Comment by kevinkamps** [comment]
I have added both options you asked for.

Changes can be found in branch [issue/1-basic-auth/master](https://github.com/kevinkamps/NomadJobUpdater/tree/issue/1-basic-auth/master). Please leave any code feedback in the [pull request](https://github.com/kevinkamps/NomadJobUpdater/pull/2).

I've tested it and it all worked but i would like to ask you to validate these changes.

I've released a newer version (1.2.0-rc1) of the application just do a `docker pull kevinkamps/nomad-job-updater:1.2.0-rc1` or download the binaries from the [release](https://github.com/kevinkamps/NomadJobUpdater/releases/tag/1.2.0-rc1) page.

For the new arguments please check the [readme](https://github.com/kevinkamps/NomadJobUpdater/tree/issue/1-basic-auth/master) page in the branch.

**Comment by jhd85** [comment]
All seems to be working, thanks!

**Comment by kevinkamps** [comment]
Thanks for validating the changes. I've released version 1.2.0.
